Solar corner cutting

hx3d opened the pull request:
rewrite of solar MPPT algorithm that compares slopes on the IP curve instead of PAO. this reduces crashing and increases power output.

I was mostly focused on implementing something that works, artifacts of previous iterations made it through. I'm sure someone that hasn't been staring at this all summer can make it more clear.

get hx3d

hx3d committed:
implemented step size and direction based on the calculated slope of the
IP curve.
on each iteration it jumps back a few steps to gather information before
returning to where it started.

hx3d committed:
better slope averaging

hx3d committed:
adjusted how many slope samples to take

hx3d committed:
trying to tune for handling long periods between cycles
currently at about 60ms withg passable performance

wjeffreys96 committed:
fix typos

wjeffreys96 committed:
Remove unused variables, unneccessary whitespace, etc

ABS is unused, no need for declaration of function if we're only using
it in this file I believe
.

wjeffreys96 committed:
Remove unecessary state member variables

As far as I can tell the NextDirection enum and the direction_up_flag
are only used in one place. Because of this, we should be able to remove
these variables and clean up a lot of the wordiness here. Also, I didn't
see anywhere where the voltage_threshold_flag variable was used so I
removed it. Lastly, when the slope_error is zero, previously the
function would multiply VREF_NEGATIVE by 0 and then return 0. I added a
final else branch for if the slope_error is 0.

wjeffreys96 committed:
housecleaning

hx3d committed:
re-ordered time_helper to avoid implicit declaration

hx3d committed:
started implementing corner cutting - WIP

hx3d committed:
it can do a full cycle and only crashes when extremely low!
removed the sample spacing
fixed reference slope implementation

hx3d committed:
refactoring for clarity and tuning

indentation, rename variables for clarity,
added preprocessor directives for corner cutting.
changed tuning params for performance.

hx3d committed:
un-indented #if directives around CC_ENABLE for style

hx3d committed:
made initialization far more reliable by changing I_ADJ_INITIAL from
1500000 to 1450000

hx3d committed:
started implementing DL.
seperated the #if direvtives for CC and DL.
created a combined directive called Intensity Estimation that is
enabled whenever either CC or DL are enabled, to handle the ring buffer
they both need

hx3d committed:
refactored solar.c so the code required by both CC and DL is under a new
flag called IE(Intensity Estimation) which handles finding the IP slope.
is dynamically enabled when either CC or DL are enabled.

hx3d committed:
consolidated divisions, made timing more consistant, renamed variables
and functions for clarity, removed old comments, removed mentions to PAO
because this isn't PAO anymore.

hx3d committed:
deleted unused solar app and devlog

ThirteenFish reviewed (changes requested):
A lot of good work went into this, I'm happy to see this up for merging. I haven't commented on the MPPT algorithm itself, I think I'll need to read your writeup of that to have an informed take, but this batch of feedback should help with the implementation.

hx3d committed:
changes and cleanup for PR into master

simplified and clarified find_ip_slope.
renamed this_iadj and old_iadj for clarity.
changed timestamp logic for accuracy.
made PT slope averaging faster.
float casting has been consolidated to improve speed.
removed unused sections and other refactoring for clarity

hx3d commented on src/f0/app_solar/source/solar.c:
```diff
@@ -69,18 +130,35 @@ typedef struct {
  * @param[in] uV        output volts in uV (microVolts).
  */
+
+int32_t time_helper(MpptState *state) {
+    int32_t time_mS = TIME_I2MS(chVTGetSystemTime());
+    int32_t rtn = time_mS - state->last_time_mS;
+    state->last_time_mS = time_mS;
+    return rtn;
+}
+
+void print_state(MpptState *state) {
+    dbgprintf("shunt uV: %10d | bus mV: %10d | curr uA: %10d | power mW: %10d | iadj uV: %d | threshold flag: %d | cycle time mS: %d\r\n",
+      state->sample.shunt_uV,
+      state->sample.voltage_mV,
+      state->sample.current_uA,
+      state->sample.power_mW,
+      state->iadj_uV,
+      time_helper(state));
+}
```
I was going to remove this for the PR but they are just too useful in troubleshooting this program. I think having them there for debugging will make life easier for future maintainers.

hx3d committed:
moved main sleep and changed time calculation
to increase consistancy and make OD values more accurate
